flight/Modules/System/systemmod.c

```diff
@@ -602,20 +665,216 @@ void vApplicationIdleHook(void)
+			objEntry->evInfo.ev.event == ev->event)
+		{
+			// Already registered, do nothing
+			PIOS_Recursive_Mutex_Unlock(mutex);
+			return -1;
+		}
+	}
+    // Create handle
+	objEntry = (PeriodicObjectList*)PIOS_malloc_no_dma(sizeof(PeriodicObjectList));
+	if (objEntry == NULL) return -1;
+	objEntry->evInfo.ev.obj = ev->obj;
+	objEntry->evInfo.ev.instId = ev->instId;
+	objEntry->evInfo.ev.event = ev->event;
+	objEntry->evInfo.cb = cb;
+	objEntry->evInfo.queue = queue;
+    objEntry->updatePeriodMs = periodMs;
+    objEntry->timeToNextUpdateMs = randomize_int(periodMs); // avoid bunching of updates
+    // Add to list
+    LL_APPEND(objList, objEntry);
+	// Release lock
+	PIOS_Recursive_Mutex_Unlock(mutex);
+    return 0;
+}
+
+/**
+ * Update the period of a periodic event.
+ * \param[in] ev The event to be dispatched
+ * \param[in] cb The callback to be invoked or zero if none
+ * \param[in] queue The queue or zero if none
+ * \param[in] periodMs The period the event is generated
+ * \return Success (0), failure (-1)
+ */
+static int32_t eventPeriodicUpdate(UAVObjEvent* ev, UAVObjEventCallback cb, struct pios_queue *queue, uint16_t periodMs)
+{
+	PeriodicObjectList* objEntry;
+	// Get lock
+	PIOS_Recursive_Mutex_Lock(mutex, PIOS_MUTEX_TIMEOUT_MAX);
+	// Find object
+	LL_FOREACH(objList, objEntry)
+	{
+		if (objEntry->evInfo.cb == cb &&
+			objEntry->evInfo.queue == queue &&
+			objEntry->evInfo.ev.obj == ev->obj &&
+			objEntry->evInfo.ev.instId == ev->instId &&
+			objEntry->evInfo.ev.event == ev->event)
+		{
+			// Object found, update period
+			objEntry->updatePeriodMs = periodMs;
+			objEntry->timeToNextUpdateMs = randomize_int(periodMs); // avoid bunching of updates
+			// Release lock
+			PIOS_Recursive_Mutex_Unlock(mutex);
+			return 0;
+		}
+	}
+    // If this point is reached the object was not found
+	PIOS_Recursive_Mutex_Unlock(mutex);
+    return -1;
+}
+
+/* It can take this long before a "first callback" on a registration,
+ * so it's advantageous for it to not be too long.  (e.g. we don't have a
+ * mechanism to wakeup on list change */
+#define MAX_UPDATE_PERIOD_MS 350
+
+/**
+ * Handle periodic updates for all objects.
+ * \return The system time until the next update (in ms) or -1 if failed
+ */
+static uint32_t processPeriodicUpdates()
+{
+	PeriodicObjectList* objEntry;
+	int32_t timeNow;
+    int32_t offset;
+
+	// Get lock
+	PIOS_Recursive_Mutex_Lock(mutex, PIOS_MUTEX_TIMEOUT_MAX);
+
+    // Iterate through each object and update its timer, if zero then transmit object.
+    // Also calculate smallest delay to next update.
+    uint32_t now = PIOS_Thread_Systime();
+    uint32_t timeToNextUpdate = PIOS_Thread_Systime() + MAX_UPDATE_PERIOD_MS;
+    LL_FOREACH(objList, objEntry)
+    {
+        // If object is configured for periodic updates
+        if (objEntry->updatePeriodMs > 0)
+        {
+            // Check if time for the next update
+        	timeNow = PIOS_Thread_Systime();
```

Comment:
Could you do it selectively on lines that are already changing anyway? It's not a perfect solution, but reduces the number of times we'll have to review style changes.